From 35f2062c8508326221a6d807ece3feec766349fb Mon Sep 17 00:00:00 2001
From: Jaroslav Rohel <jrohel@redhat.com>
Date: Wed, 25 Nov 2020 15:41:59 +0100
Subject: [PATCH] [context] dnf_keyring_add_public_keys(): not generate error
 if dir problem

The function adds all public keys from "/etc/pki/rpm-gpg" (installed public
keys) to the RPM keyring.

Before patch:
Empty directory was ignored. But if the directory could not be opened
(does not exist, it is a file, another error ...) an error has occurred.

Changes:
A non-existent directory is ignored too. If the path exists but
cannot be opened by the process as a directory, warning is reported.

Closes: #1097
Approved by: Conan-Kudo
---
 libdnf/dnf-keyring.cpp | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/libdnf/dnf-keyring.cpp b/libdnf/dnf-keyring.cpp
index 6797b119..eec58c69 100644
--- a/libdnf/dnf-keyring.cpp
+++ b/libdnf/dnf-keyring.cpp
@@ -189,11 +189,17 @@ dnf_keyring_add_public_keys(rpmKeyring keyring, GError **error) try
     const gchar *gpg_dir = "/etc/pki/rpm-gpg";
     gboolean ret = TRUE;
     g_autoptr(GDir) dir = NULL;
+    GError *localError = NULL;
 
     /* search all the public key files */
-    dir = g_dir_open(gpg_dir, 0, error);
-    if (dir == NULL)
-        return FALSE;
+    dir = g_dir_open(gpg_dir, 0, &localError);
+    if (dir == NULL) {
+        if (localError->domain != G_FILE_ERROR || localError->code != G_FILE_ERROR_NOENT) {
+            g_warning("%s", localError->message);
+        }
+        g_error_free(localError);
+        return TRUE;
+    }
     do {
         const gchar *filename;
         g_autofree gchar *path_tmp = NULL;
@@ -201,7 +207,6 @@ dnf_keyring_add_public_keys(rpmKeyring keyring, GError **error) try
         if (filename == NULL)
             break;
         path_tmp = g_build_filename(gpg_dir, filename, NULL);
-        GError *localError = NULL;
         ret = dnf_keyring_add_public_key(keyring, path_tmp, &localError);
         if (!ret) {
             g_warning("%s", localError->message);
-- 
2.28.0

